### PR TITLE
Add couchbase metrics support

### PR DESCRIPTION
**Pull request by cbornet**
See the Couchbase documentation for information on the provided metrics: https://docs.couchbase.com/java-sdk/2.7/collecting-information-and-logging.html#metric-events

### CODE REVIEW

**Comment by cbornet** [comment]
Could someone review ?

**Comment by shakuzen** [comment]
> Could someone review ?

Sorry about the delay. I'll be reviewing assigned milestones for issues/pull requests this week, as we're trying to get 1.2 released soon. If this makes it into 1.2, it will be reviewed soon. If not, it will be in 1.3 and reviewed after the 1.2 release.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                .flatMapIterable(event -> ((NetworkLatencyMetricsEvent) event).latencies().entrySet())
+                .forEach(metricEntry ->  handleMetric(registry, metricEntry.getKey(), metricEntry.getValue()));
+    }
+
+    private void handleMetric(MeterRegistry registry,
+                              NetworkLatencyMetricsIdentifier id,
+                              LatencyMetric metric) {
+        LatencyMetric previous = latencies.put(id, metric);
+        if (previous == null) {
+            registerTimeGauge("max_latency", registry, id,
```
[the remark this answers is not in the document]
Comment (cbornet) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: OK, will do. Note that I'm not sure the current behavior is correct: it reports the max latency for the last event interval. I wonder if it shouldn't be aggregated globally. Can you tell ?

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: Aggregating can be done after-the-fact in the metrics backend with aggregating by tags. For example, Prometheus has a `max` function which you could use on this across whichever tags you want to find the max of the max based on your criteria. So it looks fine to me unless I'm missing something.

[the remark this answers is not in the document]
Comment (cbornet) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: Yes but here we get latencies over a period of time configurable in the driver. So when querying you get the latency for the last period of time, not globally. If you get events with latency.max = 10,10, 200,10 and only report the first and last values, you miss that there has been a 200ms max latency event.

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: Now I see what you mean. Good catch. We could probably reuse `TimeWindowMax` to hold the max then, which is used internally for Timers/Summaries.

[the remark this answers is not in the document]
Comment (cbornet) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: Maybe I could record it in a `Timer` directly ?

[the remark this answers is not in the document]
Comment (cbornet) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: I've thought about it again. I think it should be possible to extend NetworkLatencyMetricsCollector and DefaultCouchbaseEnvironment and record latencies in a micrometer object directly.

[the remark this answers is not in the document]
Comment (cbornet) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: Arf, no ! Can't do it... DefaultCouchbaseEnvironment is built by builder...

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: > I think it should be possible to extend NetworkLatencyMetricsCollector and DefaultCouchbaseEnvironment and record latencies in a micrometer object directly.

We would probably have to approach doing that in the Couchbase client code itself, if they don't expose a means to swap out their implementation.

[the remark this answers is not in the document]
Comment (cbornet) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: Yes that would probably be better. But it's going to be a pain to contribute as they use their own JIRA... So I'll pass...
Meanwhile I propose to max the min/max metric global. It doesn't have a lot of value but at least it is clear what it represents.

[the remark this answers is not in the document]
Comment (vmotolyzhenko) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/couchbase/CouchbaseMetrics.java]: > It doesn't have a lot of value but at least it is clear what it represents.

But you can easily expose percentiles values as time gauges similar to min/max
`  
      latencyMetric.percentiles()
                .forEach((percentile, value) -> registerTimeGauge(
                       buildPercentileName(prefix,percentile), 
                       registry, 
                       id,
                       l -> TIME_GAUGE_UNIT.convert(l.get(id).percentiles().get(percentile), 
                                                                                       l.get(id).timeUnit())));

    private String buildPercentileName(String prefix, Double value) {
        String name = String.format("%.1fthPercentile", value)
                            .replace(".0", "")
                            .replace(".", "");
        return prefix + "." + name;
    }
`

**Comment by shakuzen** [comment]
Now the [Couchbase SDK has a Micrometer integration](https://docs.couchbase.com/java-sdk/current/howtos/observability-metrics.html#micrometer-integration), which is where any enhancements to the Couchbase metrics offered should be made going forward.

**Pull request closed without merging**